lib/vm-watcher.js

```diff
@@ -0,0 +1,341 @@
+ *       ...
+ *   });
+ *
+ * in order to have their _onCreated function called whenever an change has
+ * occurred. Possible events to watch are:
+ *
+ *   VmCreated
+ *   VmModified
+ *   VmDeleted
+ *
+ * For debugging the name of the watcher which saw the change is included as the
+ * second argument to the event emitter listener.
+ *
+ *
+ * FUTURE WORK:
+ *
+ *  - Add ability to watch for actions that cn-agent is currently performing
+ *    and skip doing PUT's for those VMs in that window. This will allow us to
+ *    avoid putting intermediate states. We then would also want a way to update
+ *    our object at the same time that action completes since cn-agent will
+ *    gather the VM object and the workflow will do a PUT.
+ *
+ *  - When vminfod is available, we should be able to use that watcher
+ *    exclusively on those platforms, as vminfod already watches for all
+ *    property changes.
+ *
+ */
+
+var EventEmitter = require('events').EventEmitter;
+var util = require('util');
+
+var assert = require('assert-plus');
+var diff = require('deep-diff').diff;
+
+var FsWatcher = require('../lib/watchers/fs-watcher');
+var PeriodicWatcher = require('../lib/watchers/periodic-watcher');
+var ZoneeventWatcher = require('../lib/watchers/zoneevent-watcher');
+
+
+/*
+ * The VmWatcher will emit these events:
+ *
+ * - VmCreated:   A VM has been created
+ * - VmDestroyed: A VM has been destroyed
+ * - VmModified:  A VM has been modified
+ *
+ * The first argument passed to the listener is the UUID of the VM that incurred
+ * the event.
+ *
+ */
+function VmWatcher(opts) {
+    var self = this;
+
+    assert.object(opts, 'opts');
+    assert.object(opts.log, 'opts.log');
+    assert.optionalNumber(opts.periodicInterval, 'opts.periodicInterval');
+
+    // Yay bunyan!
+    self.log = opts.log;
+
+    // This is used to try to avoid emitting the same event twice when just
+    // noticed by different watchers. How it works is that when we see an update
+    // via the updateVm() function, the watcher passes us a list of properties
+    // it thinks should have changed. We can then update the entry for that VM
+    // in knownVms so that further updates can be skipped if they just report
+    // the same change.
+    self.knownVms = {};
+
+    self.fsWatcher = new FsWatcher({
+        log: opts.log,
+        updateVm: self.newUpdateHandler('fs')
+    });
+
+    self.zoneeventWatcher = new ZoneeventWatcher({
+        log: opts.log,
+        updateVm: self.newUpdateHandler('zoneevent')
+    });
+
+    // FUTURE: If someone only wants last_modified, we can avoid starting up the
+    // periodic watcher?
+
+    self.periodicWatcher = new PeriodicWatcher({
+        log: opts.log,
+        periodicInterval: opts.periodicInterval, // might be undefined
+        updateVm: self.newUpdateHandler('periodic')
+    });
+
+    // Initialize necessary properties from `EventEmitter` in this instance
+    EventEmitter.call(this);
```

Comment:
In general, the constructor of the parent class should be the first thing done in a class' constructor so that anything that happens in the child constructor can rely on the properties and methods of the parent class being properly initialized.